utils/ipprefix/ipprefix.go

```diff
@@ -0,0 +1,166 @@
+package ipprefix
+
+import (
+	"encoding/binary"
+	"net"
+)
+
+//IPcache is an interface which provides functionality to store ip's and do longest prefix match
+type IPcache interface {
+	Put(net.IP, int, interface{})
+	Get(net.IP, int) (interface{}, bool)
+	RunVal(func(val interface{}) interface{})
+	RunIP(net.IP, func(val interface{}) bool)
+}
+
+const (
+	ipv4MaskSize = 32 + 1
+	ipv6MaskSize = 128 + 1
+)
+
+type ipcache struct {
+	ipv4 []map[uint32]interface{}
```

Comment:
locks ?

Reply:
why is ipv4 and ipv6 both in the same struct - I believe these should be two different types that implement the same interface IPcache. That looks correct mainly because most functions then have an if then else check on v4 vs v6 checks.

Reply:
we need locks. I don't see how it can be done without two structures given that the index in the map is the mask. Otherwise the ipv5 vs v6 becomes the callers problem.

Reply:
Added locks in the new commit.